### PR TITLE
sway_mirror: implement wlr_mirror allowing mirroring of variable source on a single destination output

### PR DESCRIPTION
**Pull request by alex-courtis**
## Overview

`sway_mirror` uses `wlr_mirror` to establish a mirror session which renders a box from the buffer of one output (src) on another output (dst). There may be multiple srcs during a mirror session.

Please see the accompanying [wlroots merge request](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3421) which adds `wlr_mirror`.

sway is responsible for requesting which src contents will be rendered each frame, via `wlr_mirror_request_box`. The src and box are thus variable between frames.

sway vacates dst, removing all workspaces, containers etc. for the duration of the session.

There may be concurrent mirror sessions however they cannot share dst.

Cursor may be optionally renderered.

Mirroring is driven via the new mirror command, see [sway(5)](https://github.com/swaywm/sway/pull/6743/files#diff-5acb0f48ad139a9326ab7643a3e41621976a7a3e765acf60baaa1a01d45c49f0)

## Suggested Review Order
* `wlr_mirror.h` see [wlroots MR](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3421/diffs#4aba1d13b073bbe517f206698c7aa8de65c9b3d5)
* `mirror.h`
* `desktop/mirror.c`
* `commands/mirror.c`

## Flavours

Edit 2022/01/12: Content will be scaled to fit the width or height of dst.

~~Three flavours of selecting src are available, to be combined with the three `wlr_mirror` [scaling methods](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3421#overview).~~

**entire:** all of one fixed src

**box:** a fixed box within a fixed src

**container:** a variable box that describes a container's contents within a variable src. A blank dst will be rendered when the container is not visible.

sway_mirror_flavour

## Abnormal Session Termination Conditions
* See `wlr_mirror` [conditions](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3421#abnormal-termination-conditions)
* Container is destroyed

## Limitations
See `wlr_mirror` [limitations](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3421#limitations)
### Stop Specific Session
A specific mirror session may not be stopped via the command; all of them must be stopped.
This functionality could be achieved via some mirror session id system, possibly adding to it to get_tree.
### Spanning Floating Containers
Floating containers that span multiple displays cannot be rendered in their entirety.
Only the portion of the container on the src of the container's current workspace will be rendered.
This functionality is possible, see wlr_mirror [possibility](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3421#only-one-src-box-may-be-rendered-for-each-dst-frame)

split_container

## Testing
Valgrind was used extensively during development.
Two physical outputs were used, of differing resolutions and refresh rates.
* drm gles2 - success
* drm vulkan - fail: unable to run a compositor
* drm pixman - fail: see wlr_mirror [limitations](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3421#drmpixman-not-functional)
* wlr_wl backend - success
* wlr_x11 backend - success

### CODE REVIEW

File: include/sway/server.h [hunk 1]
```diff
@@ -98,6 +99,8 @@ struct sway_server {
 	struct wlr_xdg_activation_v1 *xdg_activation_v1;
 	struct wl_listener xdg_activation_v1_request_activate;
 
+	struct wl_list mirrors; // sway_mirror::link
```
Comment (alex-courtis) [inline include/sway/server.h:102]: Mutated by `mirror_create` and `mirror_destroy`.

Accessed by `mirror_destroy_all`.

File: sway/commands/mirror.c [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+	}
+
+	if (!mirror_create(&params)) {
+		return cmd_results_new(CMD_FAILURE, "Mirror failed to start, check logs.");
+	}
+
+	return cmd_results_new(CMD_SUCCESS, NULL);
+}
+
+struct cmd_results *cmd_mirror_start_container(int argc, char **argv) {
```
Comment (alex-courtis) [inline sway/commands/mirror.c:211]: subcommands could be used, however may be impractical as these starts share a lot.

File: sway/desktop/mirror.c [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+		}
+	}
+	return NULL;
+}
+
+/**
+ * Stop rendering on output, arranging root as thought the output has been
+ * disabled or unplugged.
+ */
+void vacate_output(struct sway_output *output) {
```
Comment (alex-courtis) [inline sway/desktop/mirror.c:34]: In depth review from a sway expert would be gratefully appreciated.

File: sway/desktop/mirror.c [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	wlr_output_layout_remove(root->output_layout, output->wlr_output);
+}
+
+/**
+ * Reclaim an output, arranging root as though it is a newly enabled output.
+ *
+ * Any "pending" changes that were blocked in apply_output_config during the
+ * mirror session will be applied.
+ */
+void reclaim_output(struct sway_output *output) {
```
Comment (alex-courtis) [inline sway/desktop/mirror.c:51]: In depth review from a sway expert would be gratefully appreciated.

File: sway/sway.5.scd [hunk 5]
```diff
@@ -209,6 +209,50 @@ set|plus|minus|toggle <amount>
 	effect on the output the window is currently on. See *sway-output*(5) for
 	further details.
 
+*mirror* start <entire|container|box> <dst_name> <full|aspect|center> ...
```
Comment (alex-courtis) [inline sway/sway.5.scd:212]: mirror could be an output subcommand, however it would end up very... long.

**Comment by alex-courtis** [comment]
Builds are [currently failing](https://builds.sr.ht/~emersion/job/665607) as they are not using the mirror branch of wlroots.

Is it possible to direct CI to use that branch? Or do we simply wait until it is merged...

**Comment by bl4ckb0ne** [comment]
Patches are usually done for both wlroots and sway, merged in wlroots and merged in sway.

File: sway/sway.5.scd [hunk 6]
```diff
@@ -209,6 +209,44 @@ set|plus|minus|toggle <amount>
...
+	The destination (_dst_) output will be vacated by sway and the mirrored
+	content displayed there.
+	Content will be scaled to fit the width or height of _dst_.
+	See *sway-output*(5) for details on specifying the output name.
+
+*mirror* start _entire_ <dst_name> <src_name> [show_cursor]
+	Mirrors the whole of the source (_src_) output on _dst_. The cursor on
+	_src_ is optionally shown.
+
+*mirror* start _box_ <dst_name> <x> <y> <width> <height> [show_cursor]
```
Comment (alex-courtis) [inline sway/sway.5.scd:223]: This flavour is of questionable value. If removed, the remaining two could be merged into something like:

`mirror start <src_name|con_id|focussed> <dst_name> [show_cursor]`

**Pull request closed without merging**